exts/virtio_block/lib.rs

```diff
@@ -0,0 +1,296 @@
+        let device_id = pci_device.config().device_id();
+        if device_id != 0x1040 + 2 && device_id != 0x1001 {
+            return;
+        }
+
+        trace!("virtio-blk-pci: found the device (over PCI)");
+        let transport = match VirtioModernPci::probe_pci(pci_device) {
+            Ok(transport) => transport,
+            Err(VirtioAttachError::InvalidVendorId) => {
+                // Not a virtio-net device.
+                return;
+            }
+            Err(err) => {
+                trace!("failed to attach a virtio-blk-pci as a modern device: {:?}, falling back to the legacy driver", err);
+                match VirtioLegacyPci::probe_pci(pci_device) {
+                    Ok(transport) => transport,
+                    Err(err) => {
+                        warn!(
+                            "failed to attach a virtio-net as a legacy device: {:?}",
+                            err
+                        );
+                        return;
+                    }
+                }
+            }
+        };
+
+        let virtio = match VirtioBlock::new(transport) {
+            Ok(virtio) => virtio,
+            Err(err) => {
+                warn!("failed to initialize virtio-blk-pci: {:?}", err);
+                return;
+            }
+        };
+
+        let device = Arc::new(SpinLock::new(virtio));
+        register_block_driver(Box::new(VirtioBlockDriver::new(device.clone())));
+        attach_irq(pci_device.config().interrupt_line(), move || {
+            device.lock().handle_irq();
+        });
+    }
+
+    fn probe_virtio_mmio(&self, mmio_device: &kerla_api::driver::VirtioMmioDevice) {
+        let mmio = mmio_device.mmio_base.as_vaddr();
+        let magic = unsafe { *mmio.as_ptr::<u32>() };
+        let virtio_version = unsafe { *mmio.add(4).as_ptr::<u32>() };
+        let device_id = unsafe { *mmio.add(8).as_ptr::<u32>() };
+
+        if magic != 0x74726976 {
+            return;
+        }
+
+        if virtio_version != 2 {
+            warn!("unsupported virtio device version: {}", virtio_version);
+            return;
+        }
+
+        // Device is a block device
+        if device_id != 2 {
+            return;
+        }
+
+        trace!("virtio-block: found the device (over MMIO)");
+
+        let transport = Arc::new(VirtioMmio::new(mmio_device.mmio_base));
+        let device = match VirtioBlock::new(transport) {
+            Ok(device) => Arc::new(SpinLock::new(device)),
+            Err(VirtioAttachError::InvalidVendorId) => {
+                // Not a virtio-block device.
+                return;
+            }
+            Err(err) => {
+                warn!("failed to attach a virtio-block: {:?}", err);
+                return;
+            }
+        };
+
+        register_block_driver(Box::new(VirtioBlockDriver::new(device.clone())));
+        attach_irq(mmio_device.irq, move || {
+            device.lock().handle_irq();
+        });
+    }
```

Comment:
If I were you, I would add the following line here:

```
device.write_to_disk(0, 0 /* the first sector */, [0x5a; 256].as_slice());
```

Then, run it on QEMU and see if the first sector has been modified:

```
$ hexdump -C vdisk.img | head
```